lib/posix/options/pthread.c

```diff
@@ -1098,9 +1093,7 @@ int pthread_join(pthread_t pthread, void **status)
 		break;
 	}
 
-	ret = k_thread_join(&t->thread, K_FOREVER);
-	/* other possibilities? */
-	__ASSERT_NO_MSG(ret == 0);
+	ret = k_thread_join(&t->thread, timeout);
```

Comment:
We need to process the return value of `k_thread_join()`.

This should work - what do you think @ycsin ?

```suggestion
	ret = k_thread_join(&t->thread, timeout);
	if (ret < 0) {
		return ETIMEDOUT;
  }
```

Reply:
Two questions:
- I understand your suggestion, that we need not `posix_thread_recycle();` in case we could not join?
- I took this as reference: https://www.man7.org/linux/man-pages/man3/pthread_timedjoin_np.3.html  In that case we should return `-EBUSY` if we did not wait at all (`try_join` for example) and `-ETIMEDOUT` if we timed out. This is almost what https://docs.zephyrproject.org/latest/kernel/services/threads/index.html#c.k_thread_join does. Except it returns `-EAGAIN`. Should we return `-EAGAIN` as `k_thread_join` does or `-ETIMEDOUT` as the reference suggests.

```suggestion
	ret = k_thread_join(&t->thread, timeout);
	if (ret < 0) {
		return ret;
        }
```
```suggestion
	ret = k_thread_join(&t->thread, timeout);
	if (ret == -EAGAIN) {
		return -ETIMEDOUT;
        } else if (ret < 0) {
		return ret;
	}        
```

Reply:
For proposal see latest update. I brought the assertion back for `-EDEADLK`.